_posts/2018-07-13-coroutines.md

```diff
@@ -0,0 +1,224 @@
+category: blog
+excerpt_separator: <!--break-->
+---
+
+By [Titus Winters](mailto:titus@google.com)
+
+The first hint of standard library design that takes advantage of coroutines
+([P1056](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1056r0.html))
+came through the Library Evolution Working Group (LEWG, responsible for design
+for the C++ standard library) during the Rapperswil meeting last week. It was …
+surprising. As much as I want coroutines in the language, the design smell here
+was disconcerting. I want to point out what LEWG saw before we commit to this
+direction irrevocably - coroutine types as currently designed are baffling.
+
+<!--break-->
+
+## Background
+
+Coroutines are an important new language feature for C++, but also a very old
+idea.  In its current incarnation, the major force behind Coroutines in C++ is
+Gor Nishanov (whom I absolutely love working with).  If you aren't familiar
+with the history, or want a refresher on how Coroutines are a generalization of
+subroutines (function calls), try watching Gor's talk from CppCon 2015:
+"[C++ Coroutines, a negative overhead abstraction.](https://www.youtube.com/watch?v=_fu0gx-xseY)"
+
+In a function call/subroutine world, the caller invokes some function, the
+function runs, and then the function returns execution control to the caller.
+We're all familiar with this. In an asynchronous programming model we often
+have a logically-related sequence of operations we want to execute (for
+instance, kicking off a sequence of async operations one by one) but cannot
+express as a single function, because we need to jump in and out of that flow
+of control. As a result, all necessary state gets bundled up and passed around
+as a continuation, to be invoked at some point in the future. This is generally
+known as
+[continuation passing](https://en.wikipedia.org/wiki/Continuation-passing_style)
+and you see it in languages/libraries/frameworks that are very
+functional-style, very asynchronous, or very callback-driven. It's powerful,
+but not nearly as clear as direct style programming.
+
+Coroutines are an attempt to provide a language-level ability to write in a
+direct style and execute in a continuation-passing style. For example, consider
+an example from [P1056](wg21.link/P1056):
+
+
+```cpp
+struct record
+  {
+    int id;
+    std::string name;
+    std::string description;
+  };
+
+  std::task<record> load_record(int id);
+  std::task<> save_record(record r);
+
+  std::task<void> modify_record()
+  {
+    record r = co_await load_record(123);
+    r.description = "Look, ma, no blocking!";
+    co_await save_record(std::move(r));
+  }
+```
+
+In this snippet we're imagining a to-be-specified type `std::task<T>` that is
+in many ways a lighter-weight `std::future<T>` - it isn't promising anything to
+do with concurrency or synchronization, and has no allocation or shared state.
+Functions that return `std::task` are coroutines - they can suspend (by invoking
```

Comment:
I use too many '-'s these days. Sorry.

Reply:
Removed one. But the - - aside seems valid.